### ISSUE TITLE
🦠 Model Request: QupKake: predict micro-pKa of organic molecules

### ISSUE THREAD

**Issue by LauraGomezjurado** [ersilia-os/ersilia#1186]
### Model Name

Predict micro-pKa of organic molecules

### Model Description

QupKake is an innovative approach that combines graph neural network (GNN) models with semiempirical quantum mechanical (QM) features to forecast the micro-pKa values of organic molecules. QM has a significant role in both identifying reaction sites and predicting micro-pKa values. Precisely predicting micro-pKa values is vital for comprehending and adjusting the acidity and basicity of organic compounds, This has significant applications in drug discovery, materials science, and environmental chemistry.

### Slug

qupkake-micro-pKa

### Tag

pKa

### Publication

https://doi.org/10.1021/acs.jctc.4c00328

### Source Code

https://github.com/hutchisonlab/QupKake

### License

CC-BY-4.0

**Comment by LauraGomezjurado** [comment]
I have encountered multiple issues populating the data when running the model when populating the data. Specifically, as I debug it shows that the d the DataLoader length is 0, meaning the DataLoader is not receiving any data from the dataset. This is why trainer.predict is returning None, as there are no batches to process, and therefore the terminal responds with a "NoneIterable" error".  I have tried using the command lines `qupkake smiles "Cc1cc(-n2ncc(=O)[nH]c2=O)ccc1C(=O)c1ccccc1Cl" ` as well as `qupkake file path/to/novartis_qupkake_pka.sdf` using the [data from the github repository ](https://github.com/hutchisonlab/QupKake/tree/main/data). I am not sure how to fix this issue of what is causing it. I am currently trying to go deeper into the `mol_dataset.py` file as it has the `MolDataset` initialization to load the data 

`SITE_PREDICTIONS: None
Error: sites_predictions is None
DataLoader length: 0
SITE_PREDICTIONS: None
Error: sites_predictions is None
No protonation/deprotonation sites were found.
Output file will not be created.`

**Comment by GemmaTuron** [comment]
Hi @LauraGomezjurado 

What exactly are you doing?
I just followed the instructions and got the result as an .sdf file in the `/data/output` folder:
```
git clone https://github.com/Shualdon/QupKake.git
cd qupkake
conda env create -f environment.yml
conda activate qupkake
pip install .
qupkake smiles "Cc1[nH]c2ccccc2c1CCNCc1ccc(CCC(=O)N=O)cc1"
```
Is this precisely what you are doing? And which platform are you running the code in? I'd recommend a Linux or a WSL

**Comment by LauraGomezjurado** [comment]
Thank you so much @GemmaTuron. I did as you suggested of using codespace and it worked perfectly fine! The model is running and I am able to receive the output file. Now I will look into how to convert the output file into a readable .csv

**Comment by GemmaTuron** [comment]
Hi @LauraGomezjurado 

Please update this issue more regularly so we can see where you are stuck at and provide feedback. Thanks!

**Comment by LauraGomezjurado** [comment]
Hello @GemmaTuron! Thank you so much! 
I have mainly looking into the code to change the output we discussed and will give you updates as soon as I feel something can actually work but in the meantime I am currently guiding myself with the ersilia book, specifically forking the new model repository based on the Ersilia Model Template, but I think that  for doing so I need the model to be approved to have the link to the new model repository placeholder

**Comment by LauraGomezjurado** [comment]
In the meantime, I inspected the SDF output file I was receiving from the model for the smiles `"Cc1[nH]c2ccccc2c1CCNCc1ccc(CCC(=O)N=O)cc1" `and saw that the pKa values are stored as molecule-level properties, rather than atom-level properties.  Looking into more detail the properties I saw were: 

idx: the atom index where the pKa value is associated.
pka_type: the type of pKa (acidic or basic).
pka: the pKa value itself.

So I made a function to parse the SDF file (one I downloaded from the model just to test) and had a CSV output I will attach here as well. Looking manually into the original SDF file it looks correct, maybe I can try start implementing this function into the actual `prediction.py` and `cli.py`  files from QupKake.

`import pandas as pd
from rdkit import Chem

def write_predictions_to_csv(sdf_file, csv_file):
    print(f"Writing predictions from {sdf_file} to {csv_file}...")

    supplier = Chem.SDMolSupplier(sdf_file)
    data = []

    for mol in supplier:
        if mol is None:
            continue
        mol_name = mol.GetProp('_Name') if mol.HasProp('_Name') else 'Unknown'
        idx_list = mol.GetProp('idx').split(',')
        pka_type_list = mol.GetProp('pka_type').split(',')
        pka_list = mol.GetProp('pka').split(',')

        for idx, pka_type, pka in zip(idx_list, pka_type_list, pka_list):
            data.append({
                'Molecule': mol_name,
                'AtomIndex': int(idx),
                'pKaType': pka_type,
                'pKa': float(pka)
            })
    if data:
        df = pd.DataFrame(data)
        df.to_csv(csv_file, index=False)
        print(f"Predicted pKa values written to {csv_file}")
    else:
        print("No pKa values found in the SDF file.")
`
<img width="460" alt="Screenshot 2024-07-10 at 6 54 53 PM" src="https://github.com/ersilia-os/ersilia/assets/102627399/b52fe51a-3cc9-4a78-a566-34827c4543a2">

**Comment by GemmaTuron** [comment]
Hi @LauraGomezjurado !

This is an excellent start for the parsing of the .sdf file. We now need to figure out a way to give the output as a single line (as, when we have multiple molecules, each molecule will be one row of the .csv file) We can chat latter today about this!

**Comment by GemmaTuron** [comment]
@miquelduranfrigola 
I am tagging you here to see if you can help figure out the best output format for this model - we get the pKa for several atoms in each molecule for acidic and basic environments as an sdf file.
@LauraGomezjurado can you check what happens if you pass a list of molecules? does it generate a single sdf file per molecule or all of them are appended to the same sdf?

**Comment by LauraGomezjurado** [comment]
Thank you so much @GemmaTuron,

I run the model initially only with a CSV file with 3 molecules to test and all molecules are appended to the sma single SDF file. It doesn't let me attach it here but I have gone manually through it and the file contains all three molecules.
I have also passed the SDF file through the function I mentioned to you earlier to write it as a CSV and it looks like this:

<img width="521" alt="Screenshot 2024-07-12 at 12 46 44 AM" src="https://github.com/user-attachments/assets/00dd6c81-af32-4f69-b738-3812d1f9ad70">

**Comment by LauraGomezjurado** [comment]
Thank you so much for your guidance @miquelduranfrigola,

To see if I understood properly, the plan is to extract from the SDF file the following statistics (bellow) and put them as columns - in a CSV? (or a different format?)

min_pka: minimum pka value
avg_pka: average pka value
max_pka: maximum pka value
num_pkas: number of pka values, i.e. number of atom positions with a pka (1, 2, 3, 4, 5...)

Please let me know if I understood correctly and thank you!

**Comment by miquelduranfrigola** [comment]
You understood correctly! Thanks for the nice summary. A CSV would be ideal.

**Comment by LauraGomezjurado** [comment]
Thank you Miquel for your help yesterday! I tested what you mentioned about using a mock pKa = 0 as the input of the function and it seemed to work, it just returned the CSV file empty, and did not return any error:

```
Processing SDF file: /content/sample_data/qupkake_output.sdf
Statistics saved to /content/sample_data/0pka_test.csv
[20:10:31] Warning: molecule is tagged as 2D, but at least one Z coordinate is not zero. Marking the mol as 3D.
```

However, I also tried testing a molecule like methane (smiles "C") to see if this could be an edge case for 0 as well and received the error:

`ValueError                                Traceback (most recent call last)
<ipython-input-2-8f149f925efb> in <cell line: 66>()
     64 sdf_file = '/content/sample_data/qupkake_output.sdf'
     65 csv_file = '/content/sample_data/0pka_test.csv'
---> 66 extract_pka_statistics_from_sdf(sdf_file, csv_file)

1 frames
<ipython-input-2-8f149f925efb> in <listcomp>(.0)
     22             idx_list = mol.GetProp('idx').split(',')
     23             pka_list = mol.GetProp('pka').split(',')
---> 24             pkas = [float(pka) for pka in pka_list]
     25 
     26             #adds the pkas list to the data dictionary. If the molecule already exists in data, it extends the list with the new pKa values

ValueError: could not convert string to float: 'tensor(9.4229)'`

It is because the output SDF of metahne, for some reason, shows the pKa value in the format `tensor(float number)`do you know why may be this happening? I can just adjust the function so that it handles this case and takes the float inside the tensor as the pKa, but I am not sure if this should be something I should look into a bit deeper or not? To give you bit more context, the SDF file looks like this:

`molecule
     RDKit          3D

  5  4  0  0  0  0  0  0  0  0999 V2000
   -0.9955   -0.8730    2.1381 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.2170   -1.4027    1.5847 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.3740   -1.5137    2.9376 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.8124   -0.6152    1.4606 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.5785    0.0397    2.5695 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  1  3  1  0
  1  4  1  0
  1  5  1  0
M  END
>  <idx>  (1) 
0

>  <pka_type>  (1) 
acidic

>  <pka>  (1) 
tensor(9.4229)

$$$$`

As you can see, the <pka> value is not a normal float but a float inside a tensor()

**Comment by GemmaTuron** [comment]
Hi @LauraGomezjurado 
I am trying to help you but it is difficult without seeing the code. I have created the model repository, which you can fork and start working on it straight away. If you incorporate your code in your fork I'll be able to have a better look, thanks :)

**Comment by GemmaTuron** [comment]
The tensor issue happens only with specific molecules, when a single molecule is passed? If you add methane in a list of more molecules, what result do you get?

**Comment by GemmaTuron** [comment]
From my end, I do not think that passing only a C makes much sense for calculating a pKa value, and I think the algorithm is simply not prepared. If you use any other molecule as an example it will work just fine. I need to see the `sdf_to_csv` function to provide further help.

**Comment by GemmaTuron** [comment]
And another note, we will need an exception for when a molecule does not have a protonation site:
```
(qupkake) gturon@pujarnol:~/github/ersilia-os/QupKake$ qupkake smiles "CC"
/home/gturon/github/ersilia-os/QupKake/qupkake/xtb-641/bin/xtb
/home/gturon/github/ersilia-os/QupKake/qupkake/cli.py:157: UserWarning: File data/output/qupkake_output.sdf already exists.
A new file will be created with a different name.
Please delete the existing file if you want to overwrite it.
  warnings.warn(
Processing...
Processing molecule: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00,  6.59it/s]
Done!
No protonation/deprotonation sites were found.
Output file will not be created.
```

**Comment by miquelduranfrigola** [comment]
Thanks @GemmaTuron - indeed, in our meeting on Monday I suggested passing a simple molecule (e.g. a 'C', or 'CCC', etc.) to deal with the case where no protonation sites are expected. From what I am seeing based on your comment, the model is not creating an output in those cases, which means we need to gracefully handle this situation. I suggest that, when this happens, we simply output `[None, None, None, 0]`. What do you think, @LauraGomezjurado and @GemmaTuron ?

**Comment by GemmaTuron** [comment]
for C it is creating an output but with this weird "Tensor" which I think is a bug on their end. I think we can manually extract the number if "tensor" appears in the result no?

**Comment by LauraGomezjurado** [comment]
Thank you so much @GemmaTuron, I will be doing the forking! In the meantime let me put here the function so that I can better explain the situation:

```
import pandas as pd
from rdkit import Chem
import numpy as np

def extract_pka_statistics_from_sdf(sdf_file, csv_file):
    print(f"Processing SDF file: {sdf_file}")

    #reads the SDF file into supplier, an iterable object containing molecules.
    supplier = Chem.SDMolSupplier(sdf_file)

    #data is initialized as an empty dictionary to store pKa values for each molecule.
    data = {}

    #iterate through each molecule in supplier
    for mol in supplier:
        if mol is None:
            continue
        mol_name = mol.GetProp('_Name') if mol.HasProp('_Name') else 'Unknown'

        #Checks if the molecule has the pka property
        if mol.HasProp('pka'):
            idx_list = mol.GetProp('idx').split(',')
            pka_list = mol.GetProp('pka').split(',')
            pkas = [float(pka) for pka in pka_list]

            #adds the pkas list to the data dictionary. If the molecule already exists in data, it extends the list with the new pKa values
            if mol_name not in data:
                data[mol_name] = pkas
            else:
                data[mol_name].extend(pkas)

    summary_data = []

    #iterate through the data dictionary
    '''
    If pkas list is not empty, it calculates the statistics:
      min_pka: Minimum pKa value.
      avg_pka: Average pKa value.
      max_pka: Maximum pKa value.
      num_pkas: Number of pKa values.
    '''
    for mol_name, pkas in data.items():
        if pkas:
            min_pka = np.min(pkas)
            avg_pka = np.mean(pkas)
            max_pka = np.max(pkas)
            num_pkas = len(pkas)
        else:
            min_pka = avg_pka = max_pka = num_pkas = None

        summary_data.append({
            'Molecule': mol_name,
            'min_pka': min_pka,
            'avg_pka': avg_pka,
            'max_pka': max_pka,
            'num_pkas': num_pkas
        })

    df = pd.DataFrame(summary_data)
    df.to_csv(csv_file, index=False)
    print(f"Statistics saved to {csv_file}")

extract_pka_statistics_from_sdf(sdf_file, csv_file)
```

To answer your questions, 

1) I just tested Methane with other molecules and the output was a normal float without the tensor, so you're right there was probably a bug by their end that made the molecule alone to return the tensor format. This was the normal output when I run it with other molecules (molecule 3 is C): 

```
molecule1
     RDKit          3D

  9  8  0  0  0  0  0  0  0  0999 V2000
   -1.1680    1.0461    2.9699 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.5167   -0.2309    2.4837 C   0  0  0  0  0  0  0  0  0  0  0  0
    [...]
   -0.7759   -0.4200    1.4377 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.2815   -0.9328    2.2706 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  2  3  1  0
  [...]
  2  8  1  0
  3  9  1  0
M  END
>  <idx>  (1) 
2

>  <pka_type>  (1) 
basic

>  <pka>  (1) 
3.022221

$$$$
molecule1
     RDKit          3D

  9  8  0  0  0  0  0  0  0  0999 V2000
   -1.1680    1.0461    2.9699 C   0  0  0  0  0  0  0  0  0  0  0  0
    [...]
   -0.7759   -0.4200    1.4377 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.2815   -0.9328    2.2706 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  2  3  1  0
  [...]
  2  8  1  0
  3  9  1  0
M  END
>  <idx>  (2) 
2

>  <pka_type>  (2) 
acidic

>  <pka>  (2) 
11.492812

$$$$
molecule2
     RDKit          3D

 22 21  0  0  0  0  0  0  0  0999 V2000
    0.3894    1.2661    0.6647 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.7100    0.6944   -0.7136 C   0  0  0  0  0  0  0  0  0  0  0  0
   [...]
    2.8114   -3.2716   -0.0066 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  [...]
  7 21  1  0
  7 22  1  0
M  END
>  <idx>  (3) 
2

>  <pka_type>  (3) 
basic

>  <pka>  (3) 
10.046079

$$$$
Molecule3
     RDKit          3D

  5  4  0  0  0  0  0  0  0  0999 V2000
   -0.0074    1.5293   -0.3532 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.9314    1.9651   -0.7396 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.1735    1.8978    0.6589 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.8267    1.8133   -0.9985 H   0  0  0  0  0  0  0  0  0  0  0  0
   -0.0985    0.4411   -0.3336 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  1  3  1  0
  1  4  1  0
  1  5  1  0
M  END
>  <idx>  (4) 
0

>  <pka_type>  (4) 
acidic

>  <pka>  (4) 
9.422939

$$$$
```


2) I am guessing C does have a protonation site and that's why it does return a number, even if in the tensor format. But I also tested CCC and I did get the "No protonation/deprotonation site", and I am guessing that we can handle that case separately as @miquelduranfrigola suggested?

**Comment by GemmaTuron** [comment]
Hi  @LauraGomezjurado 

I see, we need to add two exceptions to the function. One, outside, for when the sdf file is not created (will have to skip the function) the other, inside the function, you could parse the pKas_list:

`
for p in pkas_list:
    if "tensor" in p:
        ... #remove the tensor of the string and convert number into an actual float
`

**Comment by LauraGomezjurado** [comment]
Sounds good @GemmaTuron! 

Will be implemting the parsing for the "tensor" case.

A question regarding when  the sdf file is not created, do you mean when there is no protonation/deprotonation site, right?
Becasue, in that case, we may still want to run the function for example if it is with other molecules that do have protonation site, and we could do what @miquelduranfrigola suggested of putting `none`? I am just thiking on how that output would look in the CSV output if a file and not a single molecule is the input.

**Comment by GemmaTuron** [comment]
Yes indeed, if the molecule does not have pKa we should add a None and continue - if there is no sdf file at all (the user only passed a single molecule for example) - we should have an exception before running the function because it will crash

**Comment by LauraGomezjurado** [comment]
Perfect! Will work on this and on forking it with ersilia, thank you so much @GemmaTuron

**Comment by LauraGomezjurado** [comment]
I was previously downloading the SDF output file from codespace and then running the function on my own machine, but I worked on integrating the function inside of the ` cli.py ` file and modifying the `predict.py` as well so that I could make the adjustments we discussed directly in the model. So now it's integrated in the model with the modifications for the "tensor" case and the case when a single molecule with no protonation site is the input, and  working fine:

<img width="384" alt="Screenshot 2024-07-17 at 2 19 08 PM" src="https://github.com/user-attachments/assets/8d4ad5e7-d6ba-4b35-94a2-eb205fd9e590">
<img width="382" alt="Screenshot 2024-07-17 at 2 19 58 PM" src="https://github.com/user-attachments/assets/6e4f713c-a652-49a3-9f38-ec9b25c7b66d">


However, I am working on adjusting it when the input is an entire file and one of the molecules has no protonation site, but I wanted to fork the model first before continuing with that because I think it will be a bit more complicated, and I have a question regarding that:

Guiding myself with the Ersilia Book, I understand that in the `framework` folder, inside of the `model` folder, I should put all the necessary code to run the model, and `checkpoints` folder is for the data. Since QupKake has so many files, I wanted to make sure that for this case I would have put all the files inside of the folder `qupkake` [](https://github.com/hutchisonlab/QupKake/tree/main/qupkake) right?  and in the `checkpoints` I would be copying the `data` folder as well` https://github.com/hutchisonlab/QupKake/tree/main/data` , am I correct?

Thank you so much @GemmaTuron and @miquelduranfrigola  for the help!

**Comment by miquelduranfrigola** [comment]
Hi @LauraGomezjurado and @GemmaTuron - fantastic discussion. I fully agree with the approach.
Thanks @LauraGomezjurado for your very detailed and structured comments. They do help a lot.

**Comment by LauraGomezjurado** [comment]
Thank you so much @miquelduranfrigola for your explanation! I have also talked about it with @GemmaTuron  and I am currently working on it, will update you later today.

**Comment by GemmaTuron** [comment]
@LauraGomezjurado any advances we can revise?

**Comment by LauraGomezjurado** [comment]
Thank you so much for asking @GemmaTuron, I am still in the forking, so sorry for the delay, but hopefully in a couple hours I will be able to let you know something to revise. I am also trying to understand why when one molecule is input, the model is returning the None None None 0 output we discussed but when a file is the input, the molecule is just not shown in the CSV. Will come back to you a bit later

**Comment by LauraGomezjurado** [comment]
Hello, 
Hope you had a great weekend, as I mentioned I was working on getting more familiarized with the forking and the steps for the fetching following the Ersilia Book. I have worked on the local clone I made for my fork, so copied all files, put them on the framework and checkpoints respectively, etc.  I am currently working on the input/output adapters. I understand that Ersilia uses single-column files with a header, while qupkake receives either one unique string  to a CSV file with 2 columns or and SDF file. I understand we don't need  it to manage the SDF nor the CSV, but rather just the single string? I am wondering the same with the output adapter, so far I am returning the CSV format we discussed with` [min, avg, max, num]` and I am bit confused as on how to make this adaptation, perhaps if you have some input I will be working on this before our meeting later today. 

In the other hand, next step is to run it locally before running inside Ersilia and committing changes in GithHub. But given the issue I had with the operating system, I have been running the model only in the codespace, not locally, so should I commit changes in [the forked repository under my account](https://github.com/LauraGomezjurado/eos3wzy) and run it in Ersilia's codespace?

**Comment by miquelduranfrigola** [comment]
Hi @LauraGomezjurado 

Yes please commit push your changes to your forked repository so we can have a look. Then, when the input/output issue is resolved, we will work on the PR to the ersilia-os repo of your model.

**Comment by LauraGomezjurado** [comment]
Hello @miquelduranfrigola! Thank you, I have pushed the files with the` cli.py` file modified into my [forked repository](https://github.com/LauraGomezjurado/eos3wzy) and I am working on the `main.py` file for the input adapter.

**Comment by LauraGomezjurado** [comment]
Hello, thank you @miquelduranfrigola and @DhanshreeA for your help, I worked on the implementation you suggested and wanted to update you on a blocker I have had recently. 

Whenever I try running the model, not by installing as pip package, I receive the error bellow from relative import in` cli.py `. I have been running[ my fork](https://github.com/LauraGomezjurado/eos3wzy) in the codespace but I am receiving the same import error I receive whenever I try running QupKake by running `cli.py` instead of by using the qupakake command line 
This is the error I am talking about, plus the relative import error of importing` cli.py` in` main.py`
```
(base) root@codespaces-dee3ee:/workspaces/ersilia/eos3wzy/model/framework# bash run.sh /workspaces/ersilia/eos3wzy/model/framework /workspaces/ersilia/eos3wzy/model/checkpoints/mol_small_test.csv -o
Traceback (most recent call last):
  File "/workspaces/ersilia/eos3wzy/model/framework/code/main.py", line 5, in <module>
    from cli import main_smiles
  File "/workspaces/ersilia/eos3wzy/model/framework/code/cli.py", line 84, in <module>
    from ._version import get_versions
ImportError: attempted relative import with no known parent package
```
The way cli.py imports _versions.py is like this:
```
import argparse
import logging
import sys

from _version import get_versions

__version__ = get_versions()["version"]

del get_versions

logger = logging.getLogger(__name__)
```
And how I import cli.py in main.py both seem to be causing trouble

```
import os
import csv
import sys
from cli import main_smiles
```

Perhaps you may have some idea on why this may be occurring? Thank you so much

**Comment by miquelduranfrigola** [comment]
Thanks @LauraGomezjurado I have done a PR (miquelduranfrigola:patch-1) to your repo, suggesting a small change. Let's see if this helps (I haven't tested it myself)

**Comment by LauraGomezjurado** [comment]
Thank you @miquelduranfrigola!!

I did test your suggestion and it worked for `cli.py`, then the same error was raised for `predict.py`, `mol_dataset.py`, `featurizer.py` and I did the same modification, and it worked for all:

```
import sys
root = os.path.dirname(os.path.abspath(__file__))
sys.path.append(root)   

```
However, I received the same error for the `xtb.py` but I am not sure how to modify this line `from . import XTB_LOCATION` I will look more into that and let you know. 

I also modified the `main.py ` fil to call the `main_file(args)` function instead and mimic the args, and that's how I have been having it running

```
# Create a class to mimic the argument structure expected by main_file
class Args:
    def __init__(self, filename, smiles_col, name_col, root, output):
        self.filename = filename
        self.smiles_col = smiles_col
        self.name_col = name_col
        self.root = root
        self.output = output

# Set up arguments for batch processing
args = Args(filename=tmp_input_file, smiles_col="smile", name_col="molecule", root=root, output=output_file)

# RUN QUPKAKE
main_file(args)
```

**Comment by miquelduranfrigola** [comment]
Hi @LauraGomezjurado many thanks

The `from . import XTB_LOCATION` statement tries to import from the `__init__.py` file located in the same folder. Since `XTB_LOCATION` seems to be a global fixed variable, I would recommend that you simply copy it in your `xtb.py` file, instead of importing it. How does this sound?

**Comment by LauraGomezjurado** [comment]
hello @miquelduranfrigola, thank you! I did do what you suggested along with some other modifications to different files and it seems that the relative import issue is resolved; however, I am now receiving the same error I was receiving at the very begging when I tried running the model locally in MacOS. Will look deeper into it but I am not sure what may be causing it. I am running it in the codespace, haven't yet pushed all modifications to my fork, doing that now. 

This is the error I mentioned: 
```
Traceback (most recent call last):
  File "/workspaces/ersilia/eos3wzy/model/framework/code/main.py", line 64, in <module>
    main_file(args)
  File "/workspaces/ersilia/eos3wzy/model/framework/code/cli.py", line 281, in main_file
    run_pipeline(args)
  File "/workspaces/ersilia/eos3wzy/model/framework/code/cli.py", line 382, in run_pipeline
    run_prediction_pipeline(**vars(args))
  File "/workspaces/ersilia/eos3wzy/model/framework/code/predict.py", line 267, in run_prediction_pipeline
    prot_indices = predict_sites(dataset, prot_model)
  File "/workspaces/ersilia/eos3wzy/model/framework/code/predict.py", line 172, in predict_sites
    sites_indices = [
TypeError: 'NoneType' object is not iterable
```

**Comment by miquelduranfrigola** [comment]
Thanks @LauraGomezjurado - can you please provide me with:
- The link to your latest version of the code
- The SMILES string of one molecule that you definitely know should work
I will try to have a look and come back to you soon.

**Comment by LauraGomezjurado** [comment]
Thank you @miquelduranfrigola, the latest version is the ersilia's codespace, not sure if I can share the link but I think you may be able to access it; although I am committing all changes to [my fork](https://github.com/LauraGomezjurado/eos3wzy) now too. 
I am testing with this data: 
```
mols
CCO
CCN(CC)CC
```
Thank you!

**Comment by GemmaTuron** [comment]
Hi @LauraGomezjurado and  @miquelduranfrigola 

I have worked in this model and completed its implementation. The major changes are:
- I am using the qupkake code copied directly rather than installing it, so it is more fixed and we do not have to do all the steps in the run.sh that we ideally do not want to modify
- As it was now, even if you passed several molecules, it only processed the first one. It needs a column "name" to avoid overwriting the results
- The postprocessing file was taking the results of each mol without considering that each mol has a single representation in sdf for each pka value, thus giving not the mean but actually just the values individually
- I have refactored the output to give up to 10 (top 5 and bottom 5) values

Hope this is helpful! Do have a look and let me know if you have questions!
